File: ios/App/Auth/BuiltinAuthClient.swift
import Foundation

enum BuiltinAuthClientError: Error, LocalizedError {
  case invalidEmail
  case invalidCode
  case requestFailed(Int)
  case requestFailedWithMessage(Int, String)
  case invalidResponsePayload

  var errorDescription: String? {
    switch self {
    case .invalidEmail:
      return "Enter a valid email address."
    case .invalidCode:
      return "Enter the 6-digit code from your email."
    case let .requestFailed(statusCode):
      return "Authentication request failed (\(statusCode))."
    case let .requestFailedWithMessage(statusCode, message):
      return "Authentication request failed (\(statusCode)): \(message)"
    case .invalidResponsePayload:
      return "Authentication response could not be parsed."
    }
  }
}

/// Auth client for deployed flavors, mirroring web/core's BuiltinAuthClient:
/// talks to the kernel's own auth__request__api function (email one-time
/// codes, email + password, Google OAuth via the system browser, native
/// Sign in with Apple, anonymous guests). Plain URLSession — the app's only
/// SPM dependency stays Apollo.
@MainActor
final class BuiltinAuthClient: AuthClient {
  private let authURL: URL
  private let redirectURL: URL
  private let sessionStorage: SessionStorage
  private let urlSession: URLSession
  /// Set while a sign-in yielded an MFA challenge; verifyMfaCode consumes it.
  private var pendingMfaChallenge: String?
  /// A reset flow's new password, applied once its MFA challenge is answered.
  private var pendingPasswordReset: String?

  /// Shared with the GraphQL client so every request goes out with a
  /// fresh (refreshed-if-needed) access token.
  let sessionRefresher: BuiltinSessionRefresher

  init(
    authURL: URL,
    redirectURL: URL,
    sessionStorage: SessionStorage = UserDefaultsSessionStorage(),
    urlSession: URLSession = .shared
  ) {
    self.authURL = authURL
    self.redirectURL = redirectURL
    self.sessionStorage = sessionStorage
    self.urlSession = urlSession
    self.sessionRefresher = BuiltinSessionRefresher(
      authURL: authURL,
      sessionStorage: sessionStorage,
      urlSession: urlSession
    )
  }

  func restoreSession() async -> AuthSession? {
    await sessionRefresher.validSession()
  }

  func signInLocal() async throws -> AuthSession {
    throw LocalAuthClientError.notAvailable
  }

  func sendEmailCode(email: String) async throws {
    let trimmedEmail = try validatedEmail(email)
    _ = try await postJson(path: "otp", body: EmailRequest(email: trimmedEmail))
  }

  func verifyEmailCode(email: String, code: String) async throws -> AuthSession {
    let trimmedEmail = try validatedEmail(email)
    let trimmedCode = code.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmedCode.isEmpty else {
      throw BuiltinAuthClientError.invalidCode
    }
    let data = try await postJson(
      path: "verify",
      body: VerifyRequest(email: trimmedEmail, code: trimmedCode, type: "email")
    )
    return try adoptSignInResponse(data, fallbackEmail: trimmedEmail)
  }

  func signInWithPassword(email: String, password: String) async throws -> AuthSession {
    let trimmedEmail = try validatedEmail(email)
    let data = try await postJson(
      path: "token",
      body: PasswordGrantRequest(grantType: "password", email: trimmedEmail, password: password)
    )
    return try adoptSignInResponse(data, fallbackEmail: trimmedEmail)
  }

  func signUpWithPassword(email: String, password: String) async throws -> AuthSession? {
    let trimmedEmail = try validatedEmail(email)
    let data = try await postJson(
      path: "signup",
      body: SignUpRequest(email: trimmedEmail, password: password)
    )
    // An existing identity with MFA enabled is challenged even on signup.
    try interceptMfaChallenge(in: data)
    // With email delivery configured, signup returns requires_confirmation
    // and no session: that's success — the user confirms with the code.
    guard let payload = try? JSONDecoder().decode(TokenResponsePayload.self, from: data),
          let accessToken = payload.accessToken,
          !accessToken.isEmpty
    else {
      return nil
    }
    let session = AuthSession(
      accessToken: accessToken,
      refreshToken: payload.refreshToken,
      expiresAt: payload.expiresIn.map { Date().addingTimeInterval(TimeInterval($0)) },
      email: payload.email ?? trimmedEmail
    )
    sessionStorage.persistSession(session)
    return session
  }

  func requestPasswordReset(email: String) async throws {
    let trimmedEmail = try validatedEmail(email)
    _ = try await postJson(path: "recover", body: EmailRequest(email: trimmedEmail))
  }

  func completePasswordReset(email: String, code: String, newPassword: String) async throws -> AuthSession {
    let trimmedEmail = try validatedEmail(email)
    let trimmedCode = code.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmedCode.isEmpty else {
      throw BuiltinAuthClientError.invalidCode
    }

    // Verify the recovery code first; success hands back a fresh session for
    // the account being recovered. With MFA enabled the verify yields a
    // challenge instead: hold the new password and apply it right after
    // verifyMfaCode succeeds (mirrors web/core's BuiltinAuthClient).
    let data = try await postJson(
      path: "verify",
      body: VerifyRequest(email: trimmedEmail, code: trimmedCode, type: "recovery")
    )
    if let challenge = decodeMfaChallenge(from: data) {
      pendingMfaChallenge = challenge
      pendingPasswordReset = newPassword
      throw MfaRequiredError()
    }
    let session = try persistSession(fromTokenData: data, fallbackEmail: trimmedEmail)

    // Then set the new password with the fresh access token; the user stays
    // signed in with the session from the verify step.
    _ = try await postJson(
      path: "password",
      body: UpdatePasswordRequest(password: newPassword),
      bearerToken: session.accessToken
    )
    return session
  }

  func signInAnonymously() async throws -> AuthSession {
    let data = try await postJson(path: "anonymous", body: EmptyRequest())
    return try persistSession(fromTokenData: data, fallbackEmail: nil)
  }

  func signInWithApple(identityToken: String, fullName: String?) async throws -> AuthSession {
    let data = try await postJson(
      path: "apple/native",
      body: AppleNativeRequest(identityToken: identityToken, fullName: fullName)
    )
    return try adoptSignInResponse(data, fallbackEmail: nil)
  }

  func oauthAuthorizeURL(provider: AuthMethod) -> URL? {
    // Apple signs in natively (signInWithApple); every other OAuth provider
    // rides the system-browser redirect flow on iOS.
    guard provider.isOAuthProvider, provider != .apple else {
      return nil
    }
    guard var components = URLComponents(
      url: authURL.appending(path: "\(provider.rawValue)/start"),
      resolvingAgainstBaseURL: false
    ) else {
      return nil
    }
    components.queryItems = [
      URLQueryItem(name: "redirect_to", value: redirectURL.absoluteString)
    ]
    return components.url
  }

  func handleIncomingURL(_ url: URL) async -> AuthSession? {
    guard isExpectedAuthRedirect(url) else {
      return nil
    }
    // With MFA enabled the callback fragment carries a challenge instead of
    // tokens; the caller checks hasPendingMfaChallenge when no session came.
    if let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
       let challenge = queryValueMap(from: components.fragment)["mfa_challenge"],
       !challenge.isEmpty
    {
      pendingMfaChallenge = challenge
      return nil
    }
    guard let session = parseSession(from: url) else {
      return nil
    }
    sessionStorage.persistSession(session)
    return session
  }

  func signOut() async {
    if let refreshToken = sessionStorage.loadSession()?.refreshToken {
      // Best effort: local sign-out proceeds even if revocation fails.
      _ = try? await postJson(path: "signout", body: SignOutRequest(refreshToken: refreshToken))
    }
    sessionStorage.clearSession()
  }

  func fetchRuntimeAuthMethods() async -> [AuthMethod]? {
    // Fail-safe: the sign-in surface already rendered its build-time
    // methods; a missing or malformed runtime config must never blank it.
    struct RuntimeConfigResponse: Decodable {
      let methods: [String]?
    }
    do {
      let (data, response) = try await urlSession.data(from: authURL.appending(path: "config"))
      guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
        return nil
      }
      guard let names = try JSONDecoder().decode(RuntimeConfigResponse.self, from: data).methods else {
        return nil
      }
      var methods: [AuthMethod] = []
      for name in names {
        if let method = AuthMethod(rawValue: name), !methods.contains(method) {
          methods.append(method)
        }
      }
      return methods.isEmpty ? nil : methods
    } catch {
      return nil
    }
  }

  // MARK: - Two-factor authentication

  var hasPendingMfaChallenge: Bool {
    pendingMfaChallenge != nil
  }

  func verifyMfaCode(_ code: String) async throws -> AuthSession {
    guard let challengeToken = pendingMfaChallenge else {
      throw BuiltinAuthClientError.requestFailedWithMessage(
        400, "No sign-in is awaiting a code. Start over from the sign-in screen."
      )
    }
    let trimmedCode = code.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmedCode.isEmpty else {
      throw BuiltinAuthClientError.invalidCode
    }
    let data = try await postJson(
      path: "mfa/verify",
      body: MfaVerifyRequest(challengeToken: challengeToken, code: trimmedCode)
    )
    let session = try persistSession(fromTokenData: data, fallbackEmail: nil)
    pendingMfaChallenge = nil
    if let newPassword = pendingPasswordReset {
      pendingPasswordReset = nil
      _ = try await postJson(
        path: "password",
        body: UpdatePasswordRequest(password: newPassword),
        bearerToken: session.accessToken
      )
    }
    return session
  }

  func enrollMfa() async throws -> MfaEnrollment {
    let data = try await postJson(
      path: "mfa/enroll",
      body: EmptyRequest(),
      bearerToken: try await requireAccessToken()
    )
    guard let payload = try? JSONDecoder().decode(MfaEnrollmentPayload.self, from: data) else {
      throw BuiltinAuthClientError.invalidResponsePayload
    }
    return MfaEnrollment(secret: payload.secret, otpauthUri: payload.otpauthUri)
  }

  func confirmMfa(code: String) async throws -> [String] {
    let data = try await postJson(
      path: "mfa/confirm",
      body: MfaCodeRequest(code: code),
      bearerToken: try await requireAccessToken()
    )
    guard let payload = try? JSONDecoder().decode(MfaRecoveryCodesPayload.self, from: data) else {
      throw BuiltinAuthClientError.invalidResponsePayload
    }
    return payload.recoveryCodes
  }

  func disableMfa(code: String) async throws {
    _ = try await postJson(
      path: "mfa/disable",
      body: MfaCodeRequest(code: code),
      bearerToken: try await requireAccessToken()
    )
  }

  func fetchMfaEnabled() async -> Bool {
    // Fail-safe like fetchRuntimeAuthMethods: the Settings surface must
    // render even when the status endpoint is unreachable.
    struct MfaStatusResponse: Decodable {
      let enabled: Bool?
    }
    do {
      var request = URLRequest(url: authURL.appending(path: "mfa/status"))
      request.setValue("Bearer \(try await requireAccessToken())", forHTTPHeaderField: "Authorization")
      let (data, response) = try await urlSession.data(for: request)
      guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
        return false
      }
      return (try JSONDecoder().decode(MfaStatusResponse.self, from: data)).enabled == true
    } catch {
      return false
    }
  }

  private func requireAccessToken() async throws -> String {
    guard let session = await sessionRefresher.validSession() else {
      throw BuiltinAuthClientError.requestFailedWithMessage(
        401, "You must be signed in to manage two-factor authentication."
      )
    }
    return session.accessToken
  }

  /// Adopts a sign-in response: persists the session, or — when the backend
  /// answered with an MFA challenge — holds the challenge token and throws
  /// MfaRequiredError so the surface can switch to its challenge view.
  private func adoptSignInResponse(_ data: Data, fallbackEmail: String?) throws -> AuthSession {
    try interceptMfaChallenge(in: data)
    return try persistSession(fromTokenData: data, fallbackEmail: fallbackEmail)
  }

  private func interceptMfaChallenge(in data: Data) throws {
    if let challenge = decodeMfaChallenge(from: data) {
      pendingMfaChallenge = challenge
      throw MfaRequiredError()
    }
  }

  private func decodeMfaChallenge(from data: Data) -> String? {
    guard let payload = try? JSONDecoder().decode(MfaChallengePayload.self, from: data),
          payload.mfaRequired == true,
          let challengeToken = payload.challengeToken,
          !challengeToken.isEmpty
    else {
      return nil
    }
    return challengeToken
  }

  // MARK: - URL callback parsing

  private func isExpectedAuthRedirect(_ url: URL) -> Bool {
    guard url.scheme == redirectURL.scheme else {
      return false
    }
    if let callbackHost = redirectURL.host {
      guard url.host == callbackHost else {
        return false
      }
    }
    return url.path == redirectURL.path
  }

  /// Auth callbacks carry the session in the URL fragment.
  private func parseSession(from url: URL) -> AuthSession? {
    guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
      return nil
    }
    let fragmentValues = queryValueMap(from: components.fragment)
    let queryValues = Dictionary(uniqueKeysWithValues: (components.queryItems ?? []).map { ($0.name, $0.value ?? "") })
    guard let accessToken = fragmentValues["access_token"] ?? queryValues["access_token"], !accessToken.isEmpty else {
      return nil
    }

    let expiresAt: Date?
    if let expiresInRaw = fragmentValues["expires_in"] ?? queryValues["expires_in"], let expiresIn = Double(expiresInRaw) {
      expiresAt = Date().addingTimeInterval(expiresIn)
    } else {
      expiresAt = nil
    }

    return AuthSession(
      accessToken: accessToken,
      refreshToken: fragmentValues["refresh_token"] ?? queryValues["refresh_token"],
      expiresAt: expiresAt,
      email: queryValues["email"] ?? fragmentValues["email"]
    )
  }

  private func queryValueMap(from fragment: String?) -> [String: String] {
    guard let fragment else {
      return [:]
    }

    return Dictionary(
      uniqueKeysWithValues:
        fragment
        .split(separator: "&")
        .compactMap { pair -> (String, String)? in
          let pieces = pair.split(separator: "=", maxSplits: 1)
          guard pieces.count == 2 else { return nil }
          let key = String(pieces[0])
          let value = String(pieces[1])
            .replacingOccurrences(of: "+", with: " ")
            .removingPercentEncoding ?? ""
          return (key, value)
        }
    )
  }

  // MARK: - Transport

  private func postJson(path: String, body: some Encodable, bearerToken: String? = nil) async throws -> Data {
    var request = URLRequest(url: authURL.appending(path: path))
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    if let bearerToken {
      request.setValue("Bearer \(bearerToken)", forHTTPHeaderField: "Authorization")
    }
    request.httpBody = try JSONEncoder().encode(body)

    let (data, response) = try await urlSession.data(for: request)
    guard let httpResponse = response as? HTTPURLResponse else {
      throw BuiltinAuthClientError.requestFailed(-1)
    }
    guard 200..<300 ~= httpResponse.statusCode else {
      if let message = parseErrorMessage(data: data) {
        throw BuiltinAuthClientError.requestFailedWithMessage(httpResponse.statusCode, message)
      }
      throw BuiltinAuthClientError.requestFailed(httpResponse.statusCode)
    }
    return data
  }

  private func parseErrorMessage(data: Data) -> String? {
    guard !data.isEmpty else {
      return nil
    }
    guard let payload = try? JSONDecoder().decode(ErrorPayload.self, from: data) else {
      return nil
    }
    let trimmed = payload.error?.message?.trimmingCharacters(in: .whitespacesAndNewlines)
    return trimmed?.isEmpty == false ? trimmed : nil
  }

  private func persistSession(fromTokenData data: Data, fallbackEmail: String?) throws -> AuthSession {
    guard let payload = try? JSONDecoder().decode(TokenResponsePayload.self, from: data),
          let accessToken = payload.accessToken,
          !accessToken.isEmpty
    else {
      throw BuiltinAuthClientError.invalidResponsePayload
    }
    let session = AuthSession(
      accessToken: accessToken,
      refreshToken: payload.refreshToken,
      expiresAt: payload.expiresIn.map { Date().addingTimeInterval(TimeInterval($0)) },
      email: payload.email ?? fallbackEmail
    )
    sessionStorage.persistSession(session)
    return session
  }

  private func validatedEmail(_ email: String) throws -> String {
    let trimmedEmail = email.trimmingCharacters(in: .whitespacesAndNewlines)
    guard (try? Self.emailRegex.wholeMatch(in: trimmedEmail)) != nil else {
      throw BuiltinAuthClientError.invalidEmail
    }
    return trimmedEmail
  }

  // MARK: - Payloads

  private struct EmailRequest: Encodable {
    let email: String
  }

  private struct VerifyRequest: Encodable {
    let email: String
    let code: String
    let type: String
  }

  private struct PasswordGrantRequest: Encodable {
    let grantType: String
    let email: String
    let password: String

    enum CodingKeys: String, CodingKey {
      case grantType = "grant_type"
      case email
      case password
    }
  }

  private struct SignUpRequest: Encodable {
    let email: String
    let password: String
  }

  private struct AppleNativeRequest: Encodable {
    let identityToken: String
    let fullName: String?

    enum CodingKeys: String, CodingKey {
      case identityToken = "identity_token"
      case fullName = "full_name"
    }
  }

  private struct UpdatePasswordRequest: Encodable {
    let password: String
  }

  private struct SignOutRequest: Encodable {
    let refreshToken: String

    enum CodingKeys: String, CodingKey {
      case refreshToken = "refresh_token"
    }
  }

  private struct EmptyRequest: Encodable {}

  private struct MfaVerifyRequest: Encodable {
    let challengeToken: String
    let code: String

    enum CodingKeys: String, CodingKey {
      case challengeToken = "challenge_token"
      case code
    }
  }

  private struct MfaCodeRequest: Encodable {
    let code: String
  }

  private struct MfaChallengePayload: Decodable {
    let mfaRequired: Bool?
    let challengeToken: String?

    enum CodingKeys: String, CodingKey {
      case mfaRequired = "mfa_required"
      case challengeToken = "challenge_token"
    }
  }

  private struct MfaEnrollmentPayload: Decodable {
    let secret: String
    let otpauthUri: String

    enum CodingKeys: String, CodingKey {
      case secret
      case otpauthUri = "otpauth_uri"
    }
  }

  private struct MfaRecoveryCodesPayload: Decodable {
    let recoveryCodes: [String]

    enum CodingKeys: String, CodingKey {
      case recoveryCodes = "recovery_codes"
    }
  }

  private struct TokenResponsePayload: Decodable {
    let accessToken: String?
    let refreshToken: String?
    let expiresIn: Int?
    let email: String?

    enum CodingKeys: String, CodingKey {
      case accessToken = "access_token"
      case refreshToken = "refresh_token"
      case expiresIn = "expires_in"
      case email
    }
  }

  private struct ErrorPayload: Decodable {
    struct ErrorBody: Decodable {
      let code: String?
      let message: String?
    }

    let error: ErrorBody?
  }

  private static let emailRegex = /[^\s@]+@[^\s@]+\.[^\s@]+/
}

File: web/core/src/Auth/BuiltinAuthClient.ts
import { MfaRequiredError, type AuthClient, type MfaEnrollment } from "./AuthClient"
import { type AuthMethod, type OAuthProvider } from "./AuthMethods"
import { fetchRuntimeAuthMethodsFromUrl } from "./RuntimeAuthMethods"

export interface BuiltinAuthClientConfig {
    /** Base URL of the auth__request__api function (see deriveAuthEndpoint). */
    authUrl: string
    /** Where OAuth / magic-link flows land; defaults to the current origin. */
    redirectTo?: string
}

/**
 * The auth API endpoint is the auth__request__api function, which lives next
 * to the GraphQL function in every environment — the emulator and the
 * platform deployer treat all exports uniformly — so its URL is the GraphQL
 * URL with the trailing function name swapped. The app passes its GraphQL URL
 * (import.meta.env.VITE_GRAPHQL_URL); core never reads env directly.
 */
export function deriveAuthEndpoint(graphqlUrl: string): string {
    const endpoint = graphqlUrl.replace(/graphql__request__api\/?$/, "auth__request__api")
    if (endpoint === graphqlUrl) {
        throw new Error(
            "Could not derive the auth endpoint: the GraphQL URL does not end with " +
                "the graphql__request__api function name.",
        )
    }
    return endpoint
}

const STORAGE_KEY = "base.builtinAuthSession"
/** Refresh this many seconds before the access token actually expires. */
const EXPIRY_MARGIN_SECONDS = 60

interface StoredSession {
    accessToken: string
    refreshToken: string
    /** Unix ms when the access token expires. */
    expiresAt: number
}

interface SessionResponse {
    access_token: string
    refresh_token: string
    expires_in: number
}

/** A sign-in response is a session — or an MFA challenge to answer first. */
type SignInResponse = Partial<SessionResponse> & {
    mfa_required?: boolean
    challenge_token?: string
}

/**
 * Auth client for deployed environments (VITE_AUTH_MODE=builtin): talks to
 * the kernel's own auth__request__api function. Sessions persist in
 * localStorage and refresh transparently; OAuth and magic-link flows land
 * back on the app with the session in the URL hash fragment, which the
 * constructor picks up and strips.
 */
export class BuiltinAuthClient implements AuthClient {
    private readonly authUrl: string
    private readonly redirectTo: string | undefined
    private listeners = new Set<(token: string | null) => void>()
    private refreshPromise: Promise<StoredSession | null> | undefined
    /** Set while a sign-in yielded an MFA challenge; verifyMfaCode consumes it. */
    private pendingMfaChallenge: string | undefined
    /** A reset flow's new password, applied once its MFA challenge is answered. */
    private pendingPasswordReset: string | undefined

    constructor(config: BuiltinAuthClientConfig) {
        this.authUrl = config.authUrl.replace(/\/$/, "")
        this.redirectTo = config.redirectTo
        this.adoptSessionFromUrl()
    }

    async getToken(): Promise<string | null> {
        const session = this.readSession()
        if (session === null) {
            return null
        }
        if (Date.now() < session.expiresAt - EXPIRY_MARGIN_SECONDS * 1000) {
            return session.accessToken
        }
        const refreshed = await this.refresh(session)
        return refreshed?.accessToken ?? null
    }

    async signInLocal(_token: string): Promise<void> {
        throw new Error("Local dev sign-in is not available in builtin auth mode.")
    }

    async signInWithOAuth(provider: OAuthProvider): Promise<void> {
        // The backend exposes one start route per registered provider; the
        // AuthMethods registry and this URL shape are the whole client seam.
        const redirectTo = encodeURIComponent(this.resolveRedirectTo())
        window.location.assign(`${this.authUrl}/${provider}/start?redirect_to=${redirectTo}`)
    }

    async signInWithMagicLink(email: string): Promise<void> {
        await this.postJson("/otp", { email })
    }

    async fetchRuntimeAuthMethods(): Promise<AuthMethod[] | undefined> {
        return fetchRuntimeAuthMethodsFromUrl(this.authUrl)
    }

    async verifyEmailOtp(email: string, code: string): Promise<void> {
        const response = await this.postJson<SignInResponse>("/verify", {
            email,
            code,
            type: "email",
        })
        this.adoptSignInResponse(response)
    }

    async signInWithPassword(email: string, password: string): Promise<void> {
        const response = await this.postJson<SignInResponse>("/token", {
            grant_type: "password",
            email,
            password,
        })
        this.adoptSignInResponse(response)
    }

    async signUpWithPassword(email: string, password: string): Promise<void> {
        const result = await this.postJson<SignInResponse>("/signup", { email, password })
        // Degraded environments (no SMTP) auto-confirm and return a session
        // (or a challenge, when the email already had MFA enabled).
        if (typeof result.access_token === "string" || result.mfa_required === true) {
            this.adoptSignInResponse(result)
        }
    }

    async requestPasswordReset(email: string): Promise<void> {
        await this.postJson("/recover", { email })
    }

    async completePasswordReset(email: string, code: string, newPassword: string): Promise<void> {
        // Verifying the recovery code creates a session, so on success the
        // auth store flips to signedIn and the app redirects. With MFA
        // enabled the verify yields a challenge instead; the new password is
        // held and applied right after verifyMfaCode succeeds.
        const response = await this.postJson<SignInResponse>("/verify", {
            email,
            code,
            type: "recovery",
        })
        if (response.mfa_required === true) {
            this.pendingPasswordReset = newPassword
        }
        const session = this.adoptSignInResponse(response)
        await this.postJson("/password", { password: newPassword }, session.accessToken)
    }

    async updatePassword(newPassword: string): Promise<void> {
        // getToken refreshes transparently, so the bearer token is always live.
        const token = await this.getToken()
        if (token === null) {
            throw new Error("You must be signed in to change your password.")
        }
        await this.postJson("/password", { password: newPassword }, token)
    }

    async signInAnonymously(): Promise<void> {
        const session = await this.postJson<SessionResponse>("/anonymous", {})
        this.storeSession(session)
    }

    async verifyMfaCode(code: string): Promise<void> {
        if (this.pendingMfaChallenge === undefined) {
            throw new Error("No sign-in is awaiting a code. Start over from the sign-in page.")
        }
        const session = await this.postJson<SessionResponse>("/mfa/verify", {
            challenge_token: this.pendingMfaChallenge,
            code,
        })
        this.pendingMfaChallenge = undefined
        this.storeSession(session)
        if (this.pendingPasswordReset !== undefined) {
            const newPassword = this.pendingPasswordReset
            this.pendingPasswordReset = undefined
            await this.postJson("/password", { password: newPassword }, session.access_token)
        }
    }

    hasPendingMfaChallenge(): boolean {
        return this.pendingMfaChallenge !== undefined
    }

    async enrollMfa(): Promise<MfaEnrollment> {
        const response = await this.postJson<{ secret: string; otpauth_uri: string }>(
            "/mfa/enroll",
            {},
            await this.requireToken(),
        )
        return { secret: response.secret, otpauthUri: response.otpauth_uri }
    }

    async confirmMfa(code: string): Promise<string[]> {
        const response = await this.postJson<{ recovery_codes: string[] }>(
            "/mfa/confirm",
            { code },
            await this.requireToken(),
        )
        return response.recovery_codes
    }

    async disableMfa(code: string): Promise<void> {
        await this.postJson("/mfa/disable", { code }, await this.requireToken())
    }

    async fetchMfaEnabled(): Promise<boolean> {
        // Fail-safe like fetchRuntimeAuthMethods: the Settings surface must
        // render even when the status endpoint is unreachable.
        try {
            const token = await this.getToken()
            if (token === null) {
                return false
            }
            const response = await fetch(`${this.authUrl}/mfa/status`, {
                headers: { Authorization: `Bearer ${token}` },
            })
            if (!response.ok) {
                return false
            }
            const body = (await response.json()) as { enabled?: unknown }
            return body.enabled === true
        } catch {
            return false
        }
    }

    async signOut(): Promise<void> {
        const session = this.readSession()
        if (session !== null) {
            // Best effort: local sign-out proceeds even if revocation fails.
            try {
                await this.postJson("/signout", { refresh_token: session.refreshToken })
            } catch {
                // Ignored; the local session is cleared regardless.
            }
        }
        this.clearSession()
    }

    onAuthStateChange(callback: (token: string | null) => void): () => void {
        this.listeners.add(callback)
        return () => {
            this.listeners.delete(callback)
        }
    }

    // ---- Internals ------------------------------------------------------

    /**
     * Adopts a sign-in response: stores the session, or — when the backend
     * answered with an MFA challenge — holds the challenge token and throws
     * MfaRequiredError so the surface can switch to its challenge view.
     */
    private adoptSignInResponse(response: SignInResponse): StoredSession {
        if (response.mfa_required === true && typeof response.challenge_token === "string") {
            this.pendingMfaChallenge = response.challenge_token
            throw new MfaRequiredError()
        }
        return this.storeSession(response as SessionResponse)
    }

    /**
     * OAuth and magic-link flows land on the app with the session in the URL
     * hash (never sent to any server, never logged). Adopt and strip it.
     * With MFA enabled the fragment carries a challenge instead of tokens.
     */
    private adoptSessionFromUrl(): void {
        if (typeof window === "undefined") {
            return
        }
        const hash = window.location.hash.replace(/^#/, "")
        if (hash === "") {
            return
        }
        const params = new URLSearchParams(hash)
        const mfaChallenge = params.get("mfa_challenge")
        if (mfaChallenge !== null) {
            this.pendingMfaChallenge = mfaChallenge
            window.history.replaceState(null, "", window.location.pathname + window.location.search)
            return
        }
        const accessToken = params.get("access_token")
        const refreshToken = params.get("refresh_token")
        if (accessToken === null || refreshToken === null) {
            return
        }
        const expiresIn = Number(params.get("expires_in") ?? "3600")
        this.storeSession({
            access_token: accessToken,
            refresh_token: refreshToken,
            expires_in: expiresIn,
        })
        window.history.replaceState(null, "", window.location.pathname + window.location.search)
    }

    private async refresh(session: StoredSession): Promise<StoredSession | null> {
        // Single-flight: concurrent getToken calls share one refresh request.
        this.refreshPromise ??= (async () => {
            try {
                const refreshed = await this.postJson<SessionResponse>("/token", {
                    grant_type: "refresh_token",
                    refresh_token: session.refreshToken,
                })
                return this.storeSession(refreshed)
            } catch {
                this.clearSession()
                return null
            } finally {
                this.refreshPromise = undefined
            }
        })()
        return await this.refreshPromise
    }

    private async postJson<T = unknown>(path: string, body: unknown, bearerToken?: string): Promise<T> {
        const headers: Record<string, string> = { "Content-Type": "application/json" }
        if (bearerToken !== undefined) {
            headers.Authorization = `Bearer ${bearerToken}`
        }
        const response = await fetch(`${this.authUrl}${path}`, {
            method: "POST",
            headers,
            body: JSON.stringify(body),
        })
        const payload: unknown = await response.json().catch(() => ({}))
        if (!response.ok) {
            const message =
                typeof payload === "object" &&
                payload !== null &&
                typeof (payload as { error?: { message?: unknown } }).error?.message === "string"
                    ? (payload as { error: { message: string } }).error.message
                    : `Auth request failed (${response.status}).`
            throw new Error(message)
        }
        return payload as T
    }

    private storeSession(session: SessionResponse): StoredSession {
        const stored: StoredSession = {
            accessToken: session.access_token,
            refreshToken: session.refresh_token,
            expiresAt: Date.now() + session.expires_in * 1000,
        }
        localStorage.setItem(STORAGE_KEY, JSON.stringify(stored))
        this.notify(stored.accessToken)
        return stored
    }

    private readSession(): StoredSession | null {
        const raw = localStorage.getItem(STORAGE_KEY)
        if (raw === null) {
            return null
        }
        try {
            const parsed = JSON.parse(raw) as StoredSession
            if (typeof parsed.accessToken === "string" && typeof parsed.refreshToken === "string") {
                return parsed
            }
        } catch {
            // Fall through to clearing the malformed value.
        }
        localStorage.removeItem(STORAGE_KEY)
        return null
    }

    private clearSession(): void {
        localStorage.removeItem(STORAGE_KEY)
        this.notify(null)
    }

    private async requireToken(): Promise<string> {
        const token = await this.getToken()
        if (token === null) {
            throw new Error("You must be signed in to manage two-factor authentication.")
        }
        return token
    }

    private resolveRedirectTo(): string {
        if (this.redirectTo !== undefined && this.redirectTo !== "") {
            return this.redirectTo
        }
        return window.location.origin
    }

    private notify(token: string | null): void {
        for (const listener of this.listeners) {
            listener(token)
        }
    }
}
